File: bindings/go/builders.go
// Handwritten endpoint builders for the transport.rest Go binding.
package transportrest

import (
	"context"
	"strings"
	"time"
)

// JourneyPlace describes from/to/via places of journey queries.
type JourneyPlace struct {
	form, id, name, address string
	lat, lon                float64
}

func StopID(id string) JourneyPlace      { return JourneyPlace{form: "id", id: id} }
func PlaceName(name string) JourneyPlace { return JourneyPlace{form: "name", name: name} }
func POI(id string, lat, lon float64) JourneyPlace {
	return JourneyPlace{form: "poi", id: id, lat: lat, lon: lon}
}
func Address(lat, lon float64, address string) JourneyPlace {
	return JourneyPlace{form: "address", address: address, lat: lat, lon: lon}
}

func (p JourneyPlace) encode(prefix string) []queryParam {
	switch p.form {
	case "id":
		return []queryParam{{prefix, p.id}}
	case "name":
		return []queryParam{{prefix + ".name", p.name}}
	case "poi":
		return []queryParam{
			{prefix + ".id", p.id},
			{prefix + ".latitude", formatFloat(p.lat)},
			{prefix + ".longitude", formatFloat(p.lon)},
		}
	default:
		return []queryParam{
			{prefix + ".latitude", formatFloat(p.lat)},
			{prefix + ".longitude", formatFloat(p.lon)},
			{prefix + ".address", p.address},
		}
	}
}

func formatFloat(f float64) string {
	return strconvFormat(f)
}

// ProductSelection filters transport products (unset keys are omitted).
type ProductSelection struct{ entries []queryParam }

func (p *ProductSelection) set(key string, enabled bool) *ProductSelection {
	value := "false"
	if enabled {
		value = "true"
	}
	for i := range p.entries {
		if p.entries[i].key == key {
			p.entries[i].value = value
			return p
		}
	}
	p.entries = append(p.entries, queryParam{key, value})
	return p
}

func (p *ProductSelection) NationalExpress(v bool) *ProductSelection {
	return p.set("nationalExpress", v)
}
func (p *ProductSelection) National(v bool) *ProductSelection { return p.set("national", v) }
func (p *ProductSelection) RegionalExpress(v bool) *ProductSelection {
	return p.set("regionalExpress", v)
}
func (p *ProductSelection) Regional(v bool) *ProductSelection { return p.set("regional", v) }
func (p *ProductSelection) Suburban(v bool) *ProductSelection { return p.set("suburban", v) }
func (p *ProductSelection) Subway(v bool) *ProductSelection   { return p.set("subway", v) }
func (p *ProductSelection) Tram(v bool) *ProductSelection     { return p.set("tram", v) }
func (p *ProductSelection) Bus(v bool) *ProductSelection      { return p.set("bus", v) }
func (p *ProductSelection) Ferry(v bool) *ProductSelection    { return p.set("ferry", v) }
func (p *ProductSelection) Taxi(v bool) *ProductSelection     { return p.set("taxi", v) }
func (p *ProductSelection) Express(v bool) *ProductSelection  { return p.set("express", v) }

// Locations starts a locations search.
func (c *Client) Locations(ctx context.Context) *LocationsBuilder {
	return &LocationsBuilder{client: c, ctx: ctx}
}

type LocationsBuilder struct {
	client *Client
	ctx    context.Context
	params []queryParam
	query  string
}

func (b *LocationsBuilder) Query(q string) *LocationsBuilder { b.query = q; return b }
func (b *LocationsBuilder) Fuzzy(v bool) *LocationsBuilder {
	b.params = append(b.params, queryParam{"fuzzy", formatBool(v)})
	return b
}
func (b *LocationsBuilder) Results(n int) *LocationsBuilder {
	b.params = append(b.params, queryParam{"results", intStr(n)})
	return b
}
func (b *LocationsBuilder) Stops(v bool) *LocationsBuilder {
	b.params = append(b.params, queryParam{"stops", formatBool(v)})
	return b
}
func (b *LocationsBuilder) Addresses(v bool) *LocationsBuilder {
	b.params = append(b.params, queryParam{"addresses", formatBool(v)})
	return b
}
func (b *LocationsBuilder) POI(v bool) *LocationsBuilder {
	b.params = append(b.params, queryParam{"poi", formatBool(v)})
	return b
}
func (b *LocationsBuilder) LinesOfStops(v bool) *LocationsBuilder {
	b.params = append(b.params, queryParam{"linesOfStops", formatBool(v)})
	return b
}
func (b *LocationsBuilder) Language(l string) *LocationsBuilder {
	b.params = append(b.params, queryParam{"language", l})
	return b
}

// Get executes the search.
func (b *LocationsBuilder) Get() ([]LocationResult, error) {
	if strings.TrimSpace(b.query) == "" {
		return nil, &InvalidParameterError{Parameter: "query", Reason: "a non-empty search term is required"}
	}
	all := append([]queryParam{{"query", b.query}}, b.params...)
	var out []LocationResult
	if err := b.client.getJSON(b.ctx, "/locations", all, &out, ""); err != nil {
		return nil, err
	}
	return out, nil
}

func formatBool(v bool) string {
	if v {
		return "true"
	}
	return "false"
}

File: bindings/java/src/main/java/io/transportrest/Builders.java
// Handwritten endpoint builders for the transport.rest Java binding.
package io.transportrest;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static io.transportrest.Errors.InvalidParameterException;

/** A place referenced by journey queries. */
public final class JourneyPlace {
    private final String form;
    private final String id;
    private final String name;
    private final String addressText;
    private final Double latitude;
    private final Double longitude;

    private JourneyPlace(String form, String id, String name, String addressText,
                         Double latitude, Double longitude) {
        this.form = form;
        this.id = id;
        this.name = name;
        this.addressText = addressText;
        this.latitude = latitude;
        this.longitude = longitude;
    }

    public static JourneyPlace stopId(String id) { return new JourneyPlace("id", id, null, null, null, null); }
    public static JourneyPlace name(String name) { return new JourneyPlace("name", null, name, null, null, null); }
    public static JourneyPlace poi(String id, double lat, double lon) {
        return new JourneyPlace("poi", id, null, null, lat, lon);
    }
    public static JourneyPlace address(double lat, double lon, String address) {
        return new JourneyPlace("address", null, null, address, lat, lon);
    }

    void encode(String prefix, List<String[]> params) {
        switch (form) {
            case "id" -> params.add(new String[]{prefix, id});
            case "name" -> params.add(new String[]{prefix + ".name", name});
            case "poi" -> {
                params.add(new String[]{prefix + ".id", id});
                params.add(new String[]{prefix + ".latitude", TransportRestClient.num(latitude)});
                params.add(new String[]{prefix + ".longitude", TransportRestClient.num(longitude)});
            }
            default -> {
                params.add(new String[]{prefix + ".latitude", TransportRestClient.num(latitude)});
                params.add(new String[]{prefix + ".longitude", TransportRestClient.num(longitude)});
                params.add(new String[]{prefix + ".address", addressText});
            }
        }
    }

    void validate(String parameter) {
        if ("poi".equals(form) && (id == null || id.isBlank()))
            throw new InvalidParameterException(parameter + ".id", "POI id must not be empty");
        if ("address".equals(form) && (addressText == null || addressText.isBlank()))
            throw new InvalidParameterException(parameter + ".address", "address must not be empty");
    }
}

/** Product filter; unset keys are omitted. */
public final class ProductSelection {
    private final Map<String, Boolean> entries = new LinkedHashMap<>();

    public ProductSelection set(String key, boolean enabled) { entries.put(key, enabled); return this; }
    public ProductSelection bus(boolean v) { return set("bus", v); }
    public ProductSelection tram(boolean v) { return set("tram", v); }
    public ProductSelection suburban(boolean v) { return set("suburban", v); }
    public ProductSelection subway(boolean v) { return set("subway", v); }
    public ProductSelection ferry(boolean v) { return set("ferry", v); }
    public ProductSelection express(boolean v) { return set("express", v); }
    public ProductSelection regional(boolean v) { return set("regional", v); }
    public ProductSelection national(boolean v) { return set("national", v); }
    public ProductSelection nationalExpress(boolean v) { return set("nationalExpress", v); }
    public ProductSelection regionalExpress(boolean v) { return set("regionalExpress", v); }
    public ProductSelection taxi(boolean v) { return set("taxi", v); }

    void encode(List<String[]> params) {
        entries.forEach((key, enabled) -> params.add(new String[]{key, TransportRestClient.boolStr(enabled)}));
    }
}
